pkg/skaffold/build/ordered.go

```diff
@@ -35,49 +35,51 @@ type ArtifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.A
 
 // For testing
 var (
-	buffSize      = bufferedLinesPerArtifact
-	runInSequence = InSequence
+	buffSize = bufferedLinesPerArtifact
 )
 
-// InParallel builds a list of artifacts in parallel but prints the logs in sequential order.
-func InParallel(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact ArtifactBuilder, concurrency int) ([]Artifact, error) {
-	if len(artifacts) == 0 {
-		return nil, nil
-	}
-
-	if len(artifacts) == 1 || concurrency == 1 {
-		return runInSequence(ctx, out, tags, artifacts, buildArtifact)
-	}
-
-	var wg sync.WaitGroup
-	defer wg.Wait()
-
+// InOrder builds a list of artifacts in dependency order.
+// `concurrency` specifies the max number of builds that can run at any one time. If concurrency is 0, then it's set to the length of the `artifacts` slice.
+// Each artifact build runs in its own goroutine. It limits the max concurrency using a buffered channel like a semaphore.
+// At the same time, it uses the `artifactDAG` to model the artifacts dependency graph and to make each artifact build wait for its required artifacts' builds.
+func InOrder(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact ArtifactBuilder, concurrency int) ([]Artifact, error) {
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
 	results := new(sync.Map)
 	outputs := make([]chan string, len(artifacts))
 
-	if concurrency == 0 {
-		concurrency = len(artifacts)
-	}
-	sem := make(chan bool, concurrency)
+	scheduler := newScheduler(artifacts, concurrency)
+	var wg sync.WaitGroup
+	defer wg.Wait()
 
-	// Run builds in //
 	wg.Add(len(artifacts))
 	for i := range artifacts {
 		outputs[i] = make(chan string, buffSize)
 		r, w := io.Pipe()
 
-		// Run build and write output/logs to piped writer and store build result in
-		// sync.Map
-		go func(i int) {
-			sem <- true
-			runBuild(ctx, w, tags, artifacts[i], results, buildArtifact)
-			<-sem
-
+		// Create a goroutine for each element in acmSlice. Each goroutine waits on its dependencies to finish building.
+		// Because our artifacts form a DAG, at least one of the goroutines should be able to start building.
+		go func(a *latest.Artifact) {
+			// Run build and write output/logs to piped writer and store build result in sync.Map
+			scheduler.run(ctx, a,
```

Comment:
I don't think separating this out with these helper functions improves the comprehensibility.  Could you try moving most of this function and `readOutputAndWriteToChannel()` onto the `scheduler`?

Reply:
> 1. Add a test case for requires to make sure this works or instructions for reviewers to check it out.

there are success and failure test cases in `scheduler_test.go`

> 2. I feel, the `waitForDependencies` should be called in a forever loop. Not sure how it is monitoring if all dependent artifacts are build.

it tries reading from every channel that represents each of its dependencies. the single for loop over the dependencies slice will complete only when all of these channels are closed. So it doesn't need a forever loop.